services/supervisor/src/multisig.rs

```diff
@@ -81,6 +81,44 @@ where
         self.index.put(id, confirmations);
         len
     }
+
+    /// Updates the stored confirmations to be an intersection of the set
+    /// with current confirmations, and set of the actual validator keys.
+    ///
+    /// This method is intended to be called before comparing the amount of
+    /// confirmations and amount of validators, so confirmations of nodes which
+    /// are not validators anymore won't be taken into account.
+    fn intersect(&mut self, id: &V, validator_keys: BTreeSet<PublicKey>) {
+        let mut confirmations = self.index.get(id).unwrap_or_default();
+
+        confirmations.0 = confirmations
+            .0
+            .intersection(&validator_keys)
+            .cloned()
+            .collect();
+
+        self.index.put(id, confirmations);
+    }
+
+    /// Calculates the intersection of current confirmations and actual list of
+    /// validators.
+    ///
+    /// Returns `true` if all validators confirmed the item, and `false` otherwise.
+    ///
+    /// This method updates the list of confirmation, leaving confirmations only from
+    /// the actual validators.
+    pub fn intersect_with_validators(
+        &mut self,
+        id: &V,
+        validator_keys: impl IntoIterator<Item = PublicKey>,
+    ) -> bool {
+        let validator_keys: BTreeSet<PublicKey> = validator_keys.into_iter().collect();
+        let validators_amount = validator_keys.len();
+
+        self.intersect(id, validator_keys);
+
+        self.confirmations(id) == validators_amount
```

Comment:
Nit: If the length was returned from `intersect`, the additional access to DB could be avoided.